### DISCUSSION TITLE
Is msgpackr deterministic?

### DISCUSSION THREAD

**Discussion by Llorx** [kriszyp/msgpackr#51] [Q&A] [answered]
I've found that `V8.serialize()` is not deterministic. You can serialize the same object, but created at different times, and you can receive different outputs because of how V8 internals work: https://github.com/nodejs/help/issues/2448

I'm thinking a way to serialize an object and receive the same output, no matter if I pushed/spliced/whatever the arrays, Maps and Sets. Is this msgpackr case?

Thank you.

**Comment by kriszyp** [comment, marked as answer]
Yes, msgpackr is deterministic: ordering of serialization follows the natural iterative order of objects, maps, and sets (order of insertion).

**Reply by Llorx** [reply]
To be clear, I don't mean ordering of serialization, but internal object representation. In V8, two very same exact arrays could lead to different representation because one was just created and the other had a push and a pop. I see that msgpackr has native addons so I don't know how it works with memory.

**Comment by kriszyp** [comment]
Yes, all msgpackr serialization should be deterministic, including arrays (exactly identical arrays should serialize to identical binary output). The native addon is only used for decoding/deserialization.
